test/functional/zmq_test.py

```diff
@@ -24,130 +43,79 @@ def setup_nodes(self):
         except ImportError:
             raise SkipTest("python3-zmq module not available.")
 
-        # Check that bitcoin has been built with ZMQ enabled
+        # Check that bitcoin has been built with ZMQ enabled.
         config = configparser.ConfigParser()
         if not self.options.configfile:
-            self.options.configfile = os.path.dirname(__file__) + "/../config.ini"
+            self.options.configfile = os.path.abspath(os.path.join(os.path.dirname(__file__), "../config.ini"))
         config.read_file(open(self.options.configfile))
 
         if not config["components"].getboolean("ENABLE_ZMQ"):
             raise SkipTest("bitcoind has not been built with zmq enabled.")
 
-        self.zmqContext = zmq.Context()
-        self.zmqSubSocket = self.zmqContext.socket(zmq.SUB)
-        self.zmqSubSocket.set(zmq.RCVTIMEO, 60000)
-        self.zmqSubSocket.setsockopt(zmq.SUBSCRIBE, b"hashblock")
-        self.zmqSubSocket.setsockopt(zmq.SUBSCRIBE, b"hashtx")
-        self.zmqSubSocket.setsockopt(zmq.SUBSCRIBE, b"rawblock")
-        self.zmqSubSocket.setsockopt(zmq.SUBSCRIBE, b"rawtx")
-        ip_address = "tcp://127.0.0.1:28332"
-        self.zmqSubSocket.connect(ip_address)
-        self.extra_args = [['-zmqpubhashblock=%s' % ip_address, '-zmqpubhashtx=%s' % ip_address,
-                       '-zmqpubrawblock=%s' % ip_address, '-zmqpubrawtx=%s' % ip_address], []]
+        # Initialize ZMQ context and socket.
+        # All messages are received in the same socket which means
+        # that this test fails if the publishing order changes.
```

Comment:
The ZMQSubscriber class makes it harder for the test to work if topics are sent in a different order.

Can you take a look at and consider squashing 3668dcdd5a7a12dea0ce5fd9a1d7d17ade0f7b75, which builds on your change, but gets rid of ZMQSubscriber, shortens the test, and lets it pass regardless of ordering of messages from different topics?

Reply:
@ryanofsky Thanks for the time spent on this. For the moment we want this test to fail if the order changes, see https://github.com/bitcoin/bitcoin/pull/11439#issuecomment-334181141.

Reply:
BTW, if the order changes it's only a matter of changing the order of the `sub.receive()` calls.

Reply:
Thanks for the input @ryanofsky . I personally think the `ZMQSubscriber` class is clearer, but it was me who pushed @promag in that direction initially, so we should see what other reviewers think.

Reply:
Comment isn't accurate. Should say "test hangs" instead of "test fails."

Reply:
Ah right, that one hangs :+1: